pkg/xds/auth/universal/auth_test.go

```diff
@@ -203,7 +206,44 @@ var _ = Describe("Authentication flow", func() {
 		err := authenticator.Authenticate(context.Background(), &dpRes, "this-is-not-valid-jwt-token")
 
 		// then
-		Expect(err).To(MatchError("could not parse token: token contains an invalid number of segments"))
+		Expect(err.Error()).To(ContainSubstring("could not parse token. kuma-cp runs with an in-memory database and its state isn't preserved between restarts." +
+			" Keep in mind that an in-memory database cannot be used with multiple instances of the control plane: token contains an invalid number of segments"))
+	})
+
+	It("should throw an error when signing key used for validation is different than for generation", func() {
+		// when
+		credential, err := issuer.Generate(ctx, builtin_issuer.DataplaneIdentity{
+			Name: "dp-1",
+			Mesh: "default",
+		}, 24*time.Hour)
+
+		// then
+		Expect(err).ToNot(HaveOccurred())
+
+		// when
+		err = authenticator.Authenticate(context.Background(), &dpRes, credential)
+
+		// then
+		Expect(err).ToNot(HaveOccurred())
+
+		// when
+		err = resManager.DeleteAll(context.Background(), &system.SecretResourceList{})
+
+		// then
+		Expect(err).ToNot(HaveOccurred())
+
+		// when
+		signingKeyManager := tokens.NewMeshedSigningKeyManager(resManager, builtin_issuer.DataplaneTokenSigningKeyPrefix("default"), "default")
+
+		// then
+		Expect(signingKeyManager.CreateDefaultSigningKey(ctx)).To(Succeed())
+
+		// when
+		err = authenticator.Authenticate(context.Background(), &dpRes, credential)
+
+		// then
+		Expect(err.Error()).To(ContainSubstring("could not parse token. kuma-cp runs with an in-memory database and its state isn't preserved between restarts." +
+			" Keep in mind that an in-memory database cannot be used with multiple instances of the control plane: crypto/rsa: verification error"))
```

Comment:
Usually, I try to avoid too many `when`s and `then`s in tests. Try to segment test on
`given/setup` - prepare environment/data. That's ok to have Expect in this section if env preparation throws error and you need to handle this
`when` - action that you are testing
`then` - expected behavior

I sometimes use `and` to separate large setups or large assertions. I also sometimes comment sections if the code is not super readable.

I'd refactor it to something like this 

```
// given
credential, err := issuer.Generate(ctx, builtin_issuer.DataplaneIdentity{
  Name: "dp-1",
  Mesh: "default",
}, 24*time.Hour)
Expect(err).ToNot(HaveOccurred())

// and a new signing key
signingKeyManager := tokens.NewMeshedSigningKeyManager(resManager, builtin_issuer.DataplaneTokenSigningKeyPrefix("default"), "default")
Expect(resManager.DeleteAll(context.Background(), &system.SecretResourceList{})).To(Succeed())
Expect(signingKeyManager.CreateDefaultSigningKey(ctx)).To(Succeed())

// when
err = authenticator.Authenticate(context.Background(), &dpRes, credential)

// then
Expect(err.Error()).To(ContainSubstring("could not parse token. kuma-cp runs with an in-memory database and its state isn't preserved between restarts." +
  " Keep in mind that an in-memory database cannot be used with multiple instances of the control plane: crypto/rsa: verification error"))
```

Reply:
Simplified structure as suggested, it makes sense for me and it is much easier to read, thanks 👍